perllib/FixMyStreet/Cobrand/Bromley.pm

```diff
@@ -1146,6 +1153,255 @@ sub garden_waste_cost {
+                    # already processed
+                    next RECORD if $cur->get_extra_metadata('dd_date') && $cur->get_extra_metadata('dd_date') eq $date;
+                }
+            }
+            if ( $p ) {
+                my $service = $self->waste_get_current_garden_sub( $p->get_extra_field_value('property_id') );
+                unless ($service) {
+                    warn "no matching service to cancel for $reference\n";
+                    next;
+                }
+                my $cancel = _duplicate_waste_report($p, 'Cancel Garden Subscription');
+                $cancel->set_extra_fields(
+                    {
+                        name => 'service_id',
+                        value => 545,
+                    },
+                    {
+                        name => 'uprn',
+                        value => $uprn,
+                    },
+                    {
+                        name => 'Container_Instruction_Action',
+                        value => $self->waste_container_actions->{remove},
+                    },
+                    {
+                        name => 'Container_Instruction_Container_Type',
+                        value => 44
+                    },
+                    {
+                        name => 'Container_Instruction_Quantity',
+                        value => $self->waste_get_sub_quantity($service),
+                    },
+                );
+                $cancel->set_extra_metadata('dd_date', $date);
+                $cancel->insert;
+                $handled = 1;
+            }
+        } else {
+            next unless $payment->{Status} eq 'Paid';
+            # we fetch the confirmed ones as well as we explicitly want to check for
+            # processed reports so we can warn on those we are missing.
+            $rs = $rs->search({ category => 'Garden Subscription' });
+            while ( my $cur = $rs->next ) {
+
+                my $type = $cur->get_extra_field_value('Subscription_Type');
+                # ad hoc payment already processed
+                if ( $payment->{YourRef} && $cur->id eq $payment->{YourRef} ) {
+                    if ( $cur->state ne 'unconfirmed' && $cur->get_extra_metadata('dd_date') eq $date ) {
+                        next RECORD;
+                    }
+                # new subscription payment already processed
+                } elsif ( !$payment->{YourRef} && $type eq $self->waste_subscription_types->{New} &&
+                           $cur->state ne 'unconfirmed' && $cur->get_extra_metadata('dd_date') eq $date
+                ) {
+                    next RECORD;
+                }
+
+                if ( $type eq $self->waste_subscription_types->{New} ) {
+                    if ( !$cur->get_extra_metadata('payerReference') ) {
+                        $cur->set_extra_metadata('payerReference', $payer);
+                    }
+                }
+                $cur->set_extra_metadata('dd_date', $date);
```

Comment:
If someone has multiple New reports (signed up, cancelled, sign up again, presumably), and we've got a payment with no YourRef, it feels like all those New reports will get here - the newest one will be unconfirmed, and the older ones will presumably not match on `dd_date`. Then they'll all have their dd_date updated. I might have misunderstood.